src/cast_core/test/SamplingLoader_test.cpp

```diff
@@ -0,0 +1,109 @@
+// Copyright 2022-present MongoDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <bsoncxx/builder/stream/document.hpp>
+#include <bsoncxx/json.hpp>
+
+#include <boost/exception/diagnostic_information.hpp>
+
+#include <yaml-cpp/yaml.h>
+
+#include <gennylib/context.hpp>
+#include <regex>
+#include <testlib/ActorHelper.hpp>
+#include <testlib/MongoTestFixture.hpp>
+#include <testlib/helpers.hpp>
+
+namespace genny {
+namespace {
+using namespace genny::testing;
+namespace bson_stream = bsoncxx::builder::stream;
+using namespace bsoncxx;
+
+TEST_CASE_METHOD(MongoTestFixture,
+                 "SamplingLoader - demo",
+                 "[standalone][single_node_replset][three_node_replset][sharded][SamplingLoader]") {
+
+    dropAllDatabases();
+    auto db = client.database("test");
+    auto collection = db.collection("sampling_loader_test");
+    // Seed some data.
+    collection.insert_many(std::vector{from_json("{\"x\": 0}"),
+                                       from_json("{\"x\": 1}"),
+                                       from_json("{\"x\": 2}"),
+                                       from_json("{\"x\": 3}"),
+                                       from_json("{\"x\": 4}")});
+
+    NodeSource nodes = NodeSource(R"(
+        SchemaVersion: 2018-07-01
+        Clients:
+          Default:
+            URI: )" + MongoTestFixture::connectionUri().to_string() +
+                                      R"(
+        Actors:
+        # In order to test something this random, we'll use a sample size equal to the collection
+        # size, and that way we can verify that every document gets re-inserted the same number of
+        # times. A smaller sample size would be non-determinisitc.
+        - Name: SamplingLoader
+          Type: SamplingLoader
+          Threads: 2
```

Comment:
Interesting! I don't see any barrier or anything like that in the run() method so I assume this race is possible. My vote would be to document it and relax the assertions if necessary.

Reply:
Oh, I was imagining a barrier for the threads within a single SamplingLoader actor. I suppose if you have multiple SamplingLoader actors things could get weird, but trying to do coordination between them feels like a layering violation since now actors would "know" about each other

Reply:
I might be wrong, but my mental picture of threads-in-same-actor vs actors-in-same-phase is that, is that the former is just a shortcut for the latter when you want to use the same actor with the same config. If this understanding is accurate, I'd say, let's not introduce any synchronization but mention the potential races in the SamplingLoader.yml so that the users of the actor aren't tempted to do cute things with tagging the re-inserted documents in some unique way and then relying on it -- because once in a while it's not gonna work.

Reply:
I see! Pointed out the races in the yaml file.